waspc/data/Generator/templates/server/src/core/auth/prismaMiddleware.js

```diff
@@ -0,0 +1,82 @@
+{{={= =}=}}
+import { hashPassword } from '../auth.js'
+import AuthError from '../AuthError.js'
+
+const EMAIL_FIELD = 'email'
+const PASSWORD_FIELD = 'password'
+
+// Ensure strong plaintext password.
+// Throws an AuthError on the first validation that fails.
+const registerUserEntityValidation = (prismaClient) => {
+  prismaClient.$use(async (params, next) => {
+    if (params.model === '{= userEntityUpper =}') {
+      if (['create', 'update', 'updateMany'].includes(params.action)) {
+        validateUserData(params.args.data, params.args, params.action)
+      } else if (params.action === 'upsert') {
+        validateUserData(params.args.create.data, params.args, 'create')
+        validateUserData(params.args.update.data, params.args, 'update')
+      }
+
+      // Remove from downstream Prisma processing to avoid "Unknown arg" error
+      delete params.args._waspSkipDefaultValidations
+      delete params.args._waspCustomValidations
+    }
+
+    return next(params)
+  })
+}
+
+// Make sure password is always hashed before storing to the database.
+const registerPasswordHashing = (prismaClient) => {
+  prismaClient.$use(async (params, next) => {
+    if (params.model === '{= userEntityUpper =}') {
+      if (['create', 'update', 'updateMany'].includes(params.action)) {
+        if (params.args.data.hasOwnProperty(PASSWORD_FIELD)) {
+          params.args.data[PASSWORD_FIELD] = await hashPassword(params.args.data[PASSWORD_FIELD])
+        }
+      } else if (params.action === 'upsert') {
+        if (params.args.create.data.hasOwnProperty(PASSWORD_FIELD)) {
+          params.args.create.data[PASSWORD_FIELD] =
+            await hashPassword(params.args.create.data[PASSWORD_FIELD])
+        }
+        if (params.args.update.data.hasOwnProperty(PASSWORD_FIELD)) {
+          params.args.update.data[PASSWORD_FIELD] =
+            await hashPassword(params.args.update.data[PASSWORD_FIELD])
+        }
+      }
+    }
+
+    return next(params)
+  })
+}
+
+export const registerAuthMiddleware = (prismaClient) => {
+  // registerUserEntityValidation must come before registerPasswordHashing
```

Comment:
```suggestion
  // NOTE(shayne): registerUserEntityValidation must come before registerPasswordHashing.
```
I used to format notes like this, but sure if we do that here @Martinsos ?

Reply:
Same! Sometimes it is not easy to say what is note and what is not hm. I guess note is someting you should read before you touch that part of the code, while normal comments are something you should read if you want to learn more. So yes, this would be a good candidate for a NOTE. We used this format with `(name)` after NOTE/TODO in Google so I also like using that, but I think it makes most sense when you think that however is reading the NOTE might need to consult the author. Here I think it is pretty clear what the NOTE is about and why so I don't think `(name)` is needed.

Reply:
Thanks for the discussion on the convention. It is helpful! I will update with `NOTE:` without `(name)` this time as suggested. 👍🏻